abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+returned to the host, `proxy_on_grpc_call_response` and/or `proxy_on_grpc_call_close` will be called
+with a unique call identifier (`return_callout_id`). The call identifier can also be used to cancel
+outstanding requests using `proxy_cancel_grpc_call` or close outstanding request using or
+`proxy_close_grpc_call`.
+
+
+### `proxy_open_grpc_stream`
+
+* params:
+  - `i32 (const char*) grpc_service_data`
+  - `i32 (size_t) grpc_service_size`
+  - `i32 (const char*) service_name_data`
+  - `i32 (size_t) service_name_size`
+  - `i32 (const char*) method_data`
+  - `i32 (size_t) method_size`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Open a connection to the gRPC service (`service_name_data`, `service_name_size`).
+
+
+### `proxy_send_grpc_call_message`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (const char*) grpc_message_data`
+  - `i32 (size_t) grpc_message_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Send gRPC message (`grpc_message_data`, `grpc_message_size`) on the existing gRPC stream
+(`callout_id`).
+
+
+### `proxy_cancel_grpc_call`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or stream (`callout_id`).
+
+
+### `proxy_close_grpc_call`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or close existing gRPC stream (`callout_id`).
+
+
+## Shared Key-Value store
+
+### `proxy_get_shared_data`
+
+* params:
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char**) return_value_data`
+  - `i32 (size_t*) return_value_size`
+  - `i32 (uint32_t*) return_cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get shared data identified by a key (`key_data`, `key_size`). The compare-and-switch value
+(`return_cas`) is returned and can be used when updating the value with `proxy_set_shared_data`.
+
+
+### `proxy_set_shared_data`
+
+* params:
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char*) value_data`
+  - `i32 (size_t) value_size`
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set shared data identified by a key (`key_data`, `key_size`) to a value (`value_data`,
+`value_size`). If compare-and-switch value (`cas`) is set, then it must match the current value in
+order to update to succeed.
+
+
+## Shared queue
+
+### `proxy_register_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Register a shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_resolve_shared_queue`
```

Comment:
I commented on this in the C++ PR, but why do we need both this function as well as `proxy_register_shared_queue`? Could we have a single function with a bool param `create_if_not_exists` or something like that? Otherwise the functionality is the same?

Reply:
Erm, sure.

Reply:
Per the comment in the other PR, if you don't want to merge the function that is fine, but I would probably rename to `lookup` to make it more clear and I would better clarify that `register` is the write side and `lookup` is the read side. cc @jplevyak